File: admin/lib/data/models/category.dart
import 'package:equatable/equatable.dart';

/// Category model for admin panel
class Category extends Equatable {
  final int id;
  final int? parentId;
  final String name;
  final String nameAr;
  final String? icon;
  final int sortOrder;
  final bool isActive;
  final String status;
  final int? createdBy;
  final String? rejectionReason;
  final DateTime createdAt;
  final List<Category> children;
  final Category? parent;

  const Category({
    required this.id,
    this.parentId,
    required this.name,
    required this.nameAr,
    this.icon,
    this.sortOrder = 0,
    this.isActive = true,
    this.status = 'active',
    this.createdBy,
    this.rejectionReason,
    required this.createdAt,
    this.children = const [],
    this.parent,
  });

  /// Create Category from JSON
  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'] as int,
      parentId: json['parentId'] as int?,
      name: json['name'] as String,
      nameAr: json['nameAr'] as String,
      icon: json['icon'] as String?,
      sortOrder: json['sortOrder'] as int? ?? 0,
      isActive: json['isActive'] as bool? ?? true,
      status: json['status'] as String? ?? 'active',
      createdBy: json['createdBy'] as int?,
      rejectionReason: json['rejectionReason'] as String?,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : DateTime.now(),
      children: json['children'] != null
          ? (json['children'] as List)
              .map((child) => Category.fromJson(child as Map<String, dynamic>))
              .toList()
          : const [],
      parent: json['parent'] != null
          ? Category.fromJson(json['parent'] as Map<String, dynamic>)
          : null,
    );
  }

  /// Convert Category to JSON for API requests
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'parentId': parentId,
      'name': name,
      'nameAr': nameAr,
      'icon': icon,
      'sortOrder': sortOrder,
      'isActive': isActive,
      'status': status,
      'createdBy': createdBy,
      'rejectionReason': rejectionReason,
    };
  }

  /// Create a copy with updated fields
  Category copyWith({
    int? id,
    int? parentId,
    String? name,
    String? nameAr,
    String? icon,
    int? sortOrder,
    bool? isActive,
    String? status,
    int? createdBy,
    String? rejectionReason,
    DateTime? createdAt,
    List<Category>? children,
    Category? parent,
  }) {
    return Category(
      id: id ?? this.id,
      parentId: parentId ?? this.parentId,
      name: name ?? this.name,
      nameAr: nameAr ?? this.nameAr,
      icon: icon ?? this.icon,
      sortOrder: sortOrder ?? this.sortOrder,
      isActive: isActive ?? this.isActive,
      status: status ?? this.status,
      createdBy: createdBy ?? this.createdBy,
      rejectionReason: rejectionReason ?? this.rejectionReason,
      createdAt: createdAt ?? this.createdAt,
      children: children ?? this.children,
      parent: parent ?? this.parent,
    );
  }

  /// Check if this is a parent category
  bool get isParentCategory => parentId == null;

  /// Check if this is a subcategory
  bool get isSubcategory => parentId != null;

  /// Check if category has subcategories
  bool get hasChildren => children.isNotEmpty;

  /// Get the display name (Arabic)
  String get displayName => nameAr;

  /// Get icon name or default
  String get iconName => icon ?? 'category';

  @override
  List<Object?> get props => [
        id,
        parentId,
        name,
        nameAr,
        icon,
        sortOrder,
        isActive,
        status,
        createdBy,
        rejectionReason,
      ];
}

/// Categories stats from admin API
class CategoriesStats {
  final int total;
  final int parents;
  final int subcategories;
  final int active;
  final int inactive;

  CategoriesStats({
    required this.total,
    required this.parents,
    required this.subcategories,
    required this.active,
    required this.inactive,
  });

  factory CategoriesStats.fromJson(Map<String, dynamic> json) {
    return CategoriesStats(
      total: json['total'] as int? ?? 0,
      parents: json['parents'] as int? ?? 0,
      subcategories: json['subcategories'] as int? ?? 0,
      active: json['active'] as int? ?? 0,
      inactive: json['inactive'] as int? ?? 0,
    );
  }
}

File: backend/src/models/Category.ts
import { DataTypes, Model, Optional } from 'sequelize';
import sequelize from '../config/database';

// Category attributes interface
export interface CategoryAttributes {
  id: number;
  parentId: number | null;
  name: string;
  nameAr: string;
  icon: string | null;
  sortOrder: number;
  isActive: boolean;
  status: 'active' | 'inactive' | 'pending' | 'rejected';
  createdBy: number | null;
  rejectionReason: string | null;
  createdAt: Date;
}

// Attributes for creation
export interface CategoryCreationAttributes
  extends Optional<
    CategoryAttributes,
    'id' | 'parentId' | 'icon' | 'sortOrder' | 'isActive' | 'status' | 'createdBy' | 'rejectionReason' | 'createdAt'
  > { }

// Category model class
class Category
  extends Model<CategoryAttributes, CategoryCreationAttributes>
  implements CategoryAttributes {
  public id!: number;
  public parentId!: number | null;
  public name!: string;
  public nameAr!: string;
  public icon!: string | null;
  public sortOrder!: number;
  public isActive!: boolean;
  public status!: 'active' | 'inactive' | 'pending' | 'rejected';
  public createdBy!: number | null;
  public rejectionReason!: string | null;
  public createdAt!: Date;

  // Virtual associations
  public readonly parent?: Category;
  public readonly children?: Category[];
  public readonly productCount?: number;

  // Helper methods
  public isParentCategory(): boolean {
    return this.parentId === null;
  }

  public isSubcategory(): boolean {
    return this.parentId !== null;
  }

  // Get localized name based on language
  public getLocalizedName(language: 'ar' | 'en' = 'ar'): string {
    return language === 'ar' ? this.nameAr : this.name;
  }

  // Format for API response
  public toJSON(): CategoryAttributes & { children?: Category[] } {
    const values = super.toJSON() as CategoryAttributes & { children?: Category[] };
    return values;
  }
}

Category.init(
  {
    id: {
      type: DataTypes.INTEGER,
      primaryKey: true,
      autoIncrement: true,
    },
    parentId: {
      type: DataTypes.INTEGER,
      allowNull: true,
      field: 'parent_id',
      references: {
        model: 'categories',
        key: 'id',
      },
    },
    name: {
      type: DataTypes.STRING(100),
      allowNull: false,
    },
    nameAr: {
      type: DataTypes.STRING(100),
      allowNull: false,
      field: 'name_ar',
    },
    icon: {
      type: DataTypes.STRING(100),
      allowNull: true,
    },
    sortOrder: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
      field: 'sort_order',
    },
    isActive: {
      type: DataTypes.BOOLEAN,
      defaultValue: true,
      field: 'is_active',
    },
    status: {
      type: DataTypes.ENUM('active', 'inactive', 'pending', 'rejected'),
      defaultValue: 'active',
      allowNull: false,
    },
    createdBy: {
      type: DataTypes.INTEGER,
      allowNull: true,
      field: 'created_by',
      references: {
        model: 'users',
        key: 'id',
      },
    },
    rejectionReason: {
      type: DataTypes.STRING(255),
      allowNull: true,
      field: 'rejection_reason',
    },
    createdAt: {
      type: DataTypes.DATE,
      field: 'created_at',
    },
  },
  {
    sequelize,
    tableName: 'categories',
    timestamps: false, // Only createdAt, no updatedAt
    underscored: true,
    indexes: [
      { fields: ['parent_id'] },
      { fields: ['is_active'] },
      { fields: ['sort_order'] },
      { fields: ['status'] },
    ],
  }
);

// Self-referencing associations for parent-child relationship
Category.belongsTo(Category, {
  as: 'parent',
  foreignKey: 'parentId',
});

Category.hasMany(Category, {
  as: 'children',
  foreignKey: 'parentId',
});

import User from './User';
Category.belongsTo(User, {
  as: 'creator',
  foreignKey: 'createdBy',
});

export default Category;

